src/provider/provider_coarse.c

```diff
@@ -0,0 +1,1405 @@
+    umf_result_t merge_success = UMF_RESULT_ERROR_UNKNOWN;
+    if (upstream_memory_provider && block->next && block->next->used == false &&
+        (block->data + block->size == block->next->data)) {
+
+        merge_success = umfMemoryProviderAllocationMerge(
+            upstream_memory_provider, block->data, block->next->data,
+            block->size + block->next->size);
+    }
+
+    if (merge_success == UMF_RESULT_SUCCESS) {
+        block_t *to_free = block->next;
+
+        if (block->next->free_list_ptr) {
+            free_blocks_rm_node(free_blocks, block->next->free_list_ptr);
+            block->next->free_list_ptr = NULL;
+        }
+
+        assert(block->data < block->next->data);
+        assert((block->data + block->size) == block->next->data);
+
+        if (block->next->next) {
+            block->next->next->prev = block;
+        }
+
+        // set neighbors
+        block->size += block->next->size;
+        block->next = block->next->next;
+
+        block_t *block_rm = ravl_tree_rm(all_blocks, to_free->data);
+        assert(block_rm == to_free);
+        (void)block_rm; // WA for unused variable error
+        umf_ba_global_free(to_free);
+    }
+
+    return block;
+}
+
+// alloc_find_origin - find the upstream allocation that contains data with given offset.
+static block_t *alloc_find_origin(block_t *alloc, size_t offset) {
+    assert(alloc);
+
+    while (offset >= alloc->size) {
+        offset -= alloc->size;
+        alloc = alloc->next;
+        assert(alloc);
+    }
+
+    return alloc;
+}
+
+typedef struct coarse_memory_provider_t {
+    umf_memory_provider_handle_t upstream_memory_provider;
+    void *init_buffer;
+
+    size_t used_size;
+    size_t alloc_size;
+
+    // upstream_alloc - tree of all blocks allocated from the upstream provider
+    struct ravl *upstream_alloc;
+
+    // all_blocks - tree of all blocks - sorted by an address of data
+    struct ravl *all_blocks;
+
+    // free_blocks - tree of free blocks - sorted by a size of data,
+    // each node contains a pointer (ravl_free_blocks_head_t)
+    // to the head of the list of free blocks of the same size
+    struct ravl *free_blocks;
+
+    struct os_mutex_t lock;
+
+    bool trace;
+    bool WA_do_not_free_upstream;
+} coarse_memory_provider_t;
+
+#ifndef NDEBUG
+// ravl_tree_get_head_block() - find the head (head->prev == NULL) of the all_blocks list.
+// It is not used in the critical path.
+static block_t *ravl_tree_get_head_block(struct ravl *rtree) {
+    // find head of blocks (head->prev == NULL)
+    block_t *block = NULL;
+    struct ravl_node *rnode = ravl_first(rtree);
+    if (!rnode) {
+        return NULL;
+    }
+
+    ravl_data_t *rdata = ravl_data(rnode);
+    assert(rdata);
+    block = rdata->value;
+    assert(block);
+    // make sure it is really the head
+    assert(block->prev == NULL);
+    return block;
+}
+
+static bool debug_check(coarse_memory_provider_t *provider) {
```

Comment:
please split this into multiple checks/functions

Reply:
Fixed in https://github.com/oneapi-src/unified-memory-framework/pull/715